{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e5f89bbd-c7ed-4e9e-bedd-06220f3fb302",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05367fbd-bd95-4150-b969-b731f0759d65",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a60080af-c5a9-4d57-b2f8-1f304dc9e048",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ACH-000768\tMDAMB231_BREAST\tMDA-MB-231\t905960\t1013\tBreast Cancer\tBasal\tFemale\tATC\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5eacc443-05bb-451f-92e3-bd37d66ec620",
   "metadata": {},
   "outputs": [],
   "source": [
    "CELLLINE='ACH-000768'\n",
    "CELLINENAME='MDA-MB-231'\n",
    "CELLTISSUE='BREAST'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5bea677d-f565-4d7f-8573-bae6d022574f",
   "metadata": {},
   "outputs": [],
   "source": [
    "CELLLINE='ACH-000837'\n",
    "CELLINENAME='NCI-H322M'\n",
    "CELLTISSUE='LUNG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d48a0143-fc64-41b1-b0d9-d1c306edec67",
   "metadata": {},
   "outputs": [],
   "source": [
    "CELLLINE='ACH-000510'\n",
    "CELLINENAME='NCI-H1299'\n",
    "CELLTISSUE='LUNG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "790429c7-569c-4abf-953f-efcd1d1ce4e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ACH-000788\tA2058_SKIN\tA2058\t906792\n",
    "CELLLINE='ACH-000788'\n",
    "CELLINENAME='A2058'\n",
    "CELLTISSUE='SKIN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5636b724-647b-4a51-9271-b9b13a48892c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ACH-001190\tSKMEL2_SKIN\tSK-MEL2\t905955\t1191\tSkin Cancer\tmalignant_melanoma\tMale\t\n",
    "CELLLINE='ACH-001190'\n",
    "CELLINENAME='SK-MEL-2'\n",
    "CELLTISSUE='SKIN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a110354f-29dd-433f-849a-76cf460e5cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "CELLLINE='ACH-000496'\n",
    "CELLINENAME='NCI-H1792'\n",
    "CELLTISSUE='LUNG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "468e8b1e-5344-4ce6-adc4-364ac7a222dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "CELLLINE='ACH-000667'\n",
    "CELLINENAME='NCI-HCC44'\n",
    "CELLTISSUE='LUNG'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea56f1a4-d429-4cfe-bfa6-ab2092f8528a",
   "metadata": {},
   "source": [
    "### Data Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9e49150d-af19-4c66-b155-efb7397454b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getRNAExpression(celline):    \n",
    "    # https://depmap.org/portal/data_page/?tab=customDownloads - Expression_Public_24Q4_subsetted.csv\n",
    "    df = pd.read_csv('data/Expression_Public_24Q4_subsetted.csv', index_col=0)\n",
    "    filtered_df = df.loc[celline]\n",
    "    transformed_df = filtered_df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={filtered_df.name: 'rnaexp'})      \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})     \n",
    "    print(\"Public RNA Expression data for 'cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2bf257c1-bf6d-4c29-9e99-f0a52f5e9ac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCN(celline):    \n",
    "    # https://depmap.org/portal/data_page/?tab=customDownloads - Copy_Number_Public_24Q4_(Log2_transformed)_subsetted.csv\n",
    "    df = pd.read_csv('data/Copy_Number_Public_24Q4_(Log2_transformed)_subsetted.csv', index_col=0)\n",
    "    filtered_df = df.loc[celline]\n",
    "    transformed_df = filtered_df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={filtered_df.name: 'copynumber'})      \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})     \n",
    "    print(\"Copy Number data for 'cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3d2a0d30-d96c-435b-8a20-87f6172bccfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCRISPRCas9(celline):    \n",
    "    # CRISPR-Cas9 gene   knockout dependency\n",
    "    # https://depmap.org/portal/data_page/?tab=customDownloads - CRISPR_(DepMap_Public_24Q4+Score,_Chronos)_subsetted.csv\n",
    "    df = pd.read_csv('data/CRISPR_(DepMap_Public_24Q4+Score,_Chronos)_subsetted.csv', index_col=0)\n",
    "    filtered_df = df.loc[celline]\n",
    "    transformed_df = filtered_df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={filtered_df.name: 'crisprcas9'})      \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})     \n",
    "    print(\"CRISPR-Cas9 gene knockout dependency data for 'cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "499c4b8f-901f-4925-b6a1-caa617ea1d9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def getRNAi(celline):    \n",
    "    # https://depmap.org/portal/data_page/?tab=customDownloads - RNAi_(Achilles+DRIVE+Marcotte,_DEMETER2)_subsetted.csv\n",
    "    df = pd.read_csv('data/RNAi_(Achilles+DRIVE+Marcotte,_DEMETER2)_subsetted.csv', index_col=0)\n",
    "    filtered_df = df.loc[celline]\n",
    "    transformed_df = filtered_df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={filtered_df.name: 'rnai'})      \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})     \n",
    "    print(\"Public RNAi data for 'cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "24cb8bf5-2848-4242-9981-b25b01846eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCCLExpression(celline):    \n",
    "    # RNAseq TPM gene expression data for all genes using RSEM. Log2 tra\n",
    "    # https://depmap.org/portal/data_page/?tab=allData&releasename=DepMap+Public+20Q1&filename=CCLE_expression_full.csv - CCLE_expression_full.csv\n",
    "    df = pd.read_csv('data/CCLE_expression_full.csv', index_col=0)\n",
    "    filtered_df = df.loc[celline]\n",
    "    transformed_df = filtered_df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={filtered_df.name: 'ccleexp'})      \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})     \n",
    "    transformed_df['gene'] = transformed_df['gene'].str.split('(').str[0].str.strip()\n",
    "    print(\"RNAseq TPM gene expression  data for 'cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4d5f4cef-11ab-4ebe-b968-f03181bd0a06",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getMethylation(celline):    \n",
    "    if celline=='ACH-000510':\n",
    "        celline='ACH-000511'\n",
    "    # RNAseq TPM gene expression data for all genes using RSEM. Log2 tra\n",
    "    # https://depmap.org/portal/data_page/?tab=customDownloads Methylation_(1kb_upstream_TSS)_subsetted_NAsdropped.csv\n",
    "    df = pd.read_csv('data/Methylation_(1kb_upstream_TSS)_subsetted_NAsdropped.csv', index_col=0)\n",
    "    filtered_df = df.loc[celline]\n",
    "    transformed_df = filtered_df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={filtered_df.name: 'methylation'})      \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})     \n",
    "    transformed_df['gene'] = transformed_df['gene'].str.split('_1_').str[0]\n",
    "    print(\"Methylation data for 'cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a93a0fa1-27f4-4823-9e16-22c57fbc75da",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getProteinGygi(celline):    \n",
    "    #The Harmonized MS CCLE Gygi protein expression dataset refers to a collection of protein expression data obtained from the CCLE\n",
    "    # which has been harmonized and generated using mass spectrometry (MS) methods, specifically following the Gygi lab's protocols.\n",
    "    # https://depmap.org/portal/data_page/?tab=customDownloads - Harmonized_MS_CCLE_Gygi_subsetted.csv\n",
    "    df = pd.read_csv('data/Harmonized_MS_CCLE_Gygi_subsetted.csv', index_col=0)\n",
    "    filtered_df = df.loc[celline]\n",
    "    transformed_df = filtered_df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={filtered_df.name: 'proteingygi'})      \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})     \n",
    "    transformed_df['gene'] = transformed_df['gene'].str.extract(r'\\((.*?)\\)', expand=False)\n",
    "    print(\"Harmonized_MS_CCLE_Gygi_subsetted  data for 'cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "09ecd04d-d8f3-43b2-be12-81fa960d51cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getProteinRPPA(celline):    \n",
    "    # TReverse Phase Protein Array (RPPA) data\n",
    "    # https://depmap.org/portal/data_page/?tab=customDownloads - Harmonized_RPPA_CCLE_subsetted.csv\n",
    "    df = pd.read_csv('data/Harmonized_RPPA_CCLE_subsetted.csv', index_col=0)\n",
    "    filtered_df = df.loc[celline]\n",
    "    transformed_df = filtered_df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={filtered_df.name: 'proteinrpna'})      \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})     \n",
    "    transformed_df['gene'] = transformed_df['gene'].str.extract(r'\\((.*?)\\)', expand=False)\n",
    "    print(\"Harmonized_RPPA_CCLE_subsetted  data for 'cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1d9cd21d-cb19-4699-b7e0-e88e58865ed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://github.com/DepMap-Analytics/CoRe/blob/master/notebooks/CoRe_Benchmarking.ipynb\n",
    "#gene dependency data derived from CRISPR screens, which are used to identify which genes are essential for the survival of cancer cells.\n",
    "#These essential genes are essentially fitness genes or CFGs.\n",
    "\n",
    "def getSangerCFGs(celline):    \n",
    "    df = pd.read_csv('data/Sanger_scaled_Behan2019.csv', index_col=0)\n",
    "    transformed_df = df.reset_index()   \n",
    "    transformed_df = transformed_df.rename(columns={'index': 'gene'})   \n",
    "    transformed_df = transformed_df.rename(columns={'gene_name': 'gene'})     \n",
    "    transformed_df = transformed_df.rename(columns={CELLINENAME: 'sangercfgs'})     \n",
    "    transformed_df = transformed_df[['gene','sangercfgs']]\n",
    "    print(\"Sanger_scaled_Behan2019 CFGs for cellline:\"+celline,transformed_df.shape)\n",
    "    print(transformed_df.head(5))\n",
    "    return transformed_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cc6f35a-5a85-4866-9ec5-91c116270199",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e3f85ffd-93c1-4aa1-8a81-88316322de79",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getGCContent():    \n",
    "    gccontent=pd.read_csv('genedata/gccontent.csv') \n",
    "    gccontent=gccontent.dropna()\n",
    "    print(gccontent.shape)\n",
    "    gccontent=gccontent[['hgnc_symbol','percentage_gene_gc_content']]\n",
    "    gccontent = gccontent.rename(columns={'hgnc_symbol': 'gene'})\n",
    "    print('gc content :',gccontent.shape)\n",
    "    print(gccontent.head())\n",
    "    return gccontent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "72d3b221-434f-493f-a398-1dc2a8446b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getGOgenes():    \n",
    "    dfgo_genes = pd.read_csv('genedata/go_genes.csv') \n",
    "    dfgo_genes=dfgo_genes.dropna()\n",
    "    dfgo_genes.shape\n",
    "    dfgo_genes=dfgo_genes[['hgnc_symbol','go_id']]\n",
    "    dfgo_genes = dfgo_genes.rename(columns={'hgnc_symbol': 'gene'})\n",
    "    print('GO Ids:',dfgo_genes.shape)\n",
    "    print(dfgo_genes.head())\n",
    "    return dfgo_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "701b9784-95ee-4b6e-a96c-ca07d4f84156",
   "metadata": {},
   "source": [
    "### Extract data now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d3cd8e33-54d7-4351-9808-3718c28547d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Public RNA Expression data for 'cellline:ACH-000667 (19153, 2)\n",
      "     gene    rnaexp\n",
      "0  TSPAN6  3.475085\n",
      "1    TNMD  0.000000\n",
      "2    DPM1  5.632850\n",
      "3   SCYL3  1.786596\n",
      "4   FIRRM  2.797013\n",
      "getRNAExpression  (19153, 2)\n",
      "Copy Number data for 'cellline:ACH-000667 (38589, 2)\n",
      "        gene  copynumber\n",
      "0       RHEB    1.228123\n",
      "1      TIPIN    0.818824\n",
      "2     OR4A47    0.957604\n",
      "3     NUP133    1.068772\n",
      "4  CPNE8-AS1    1.073707\n",
      "getCN  (19153, 3)\n",
      "CRISPR-Cas9 gene knockout dependency data for 'cellline:ACH-000667 (17916, 2)\n",
      "      gene  crisprcas9\n",
      "0     A1BG   -0.032698\n",
      "1     A1CF   -0.115469\n",
      "2      A2M    0.098039\n",
      "3    A2ML1    0.168418\n",
      "4  A3GALT2    0.029768\n",
      "getCRISPRCas9  (19153, 4)\n",
      "Public RNAi data for 'cellline:ACH-000667 (16836, 2)\n",
      "   gene      rnai\n",
      "0  A1BG -0.122757\n",
      "1  NAT2  0.154515\n",
      "2   ADA -0.250882\n",
      "3  CDH2  0.124039\n",
      "4  AKT3  0.203598\n",
      "getRNAi  (19153, 5)\n",
      "RNAseq TPM gene expression  data for 'cellline:ACH-000667 (58676, 2)\n",
      "       gene   ccleexp\n",
      "0    TSPAN6  4.221877\n",
      "1      TNMD  0.000000\n",
      "2      DPM1  5.512227\n",
      "3     SCYL3  1.794936\n",
      "4  C1orf112  2.809414\n",
      "getCCLExpression  (19175, 6)\n",
      "Methylation data for 'cellline:ACH-000667 (21337, 2)\n",
      "     gene  methylation\n",
      "0   AGBL4      0.76636\n",
      "1   AJAP1      0.88057\n",
      "2   ACOT7      0.09140\n",
      "3   ACOT7      0.00000\n",
      "4  AGTRAP      0.00000\n",
      "getMethylation  (19537, 7)\n",
      "Harmonized_MS_CCLE_Gygi_subsetted  data for 'cellline:ACH-000667 (12558, 2)\n",
      "      gene  proteingygi\n",
      "0    RBM47    -0.058328\n",
      "1    IFT56    -0.175957\n",
      "2  TSNARE1          NaN\n",
      "3  TMEM129          NaN\n",
      "4     E2F8          NaN\n",
      "getProteinGygi  (19989, 8)\n",
      "Harmonized_RPPA_CCLE_subsetted  data for 'cellline:ACH-000667 (144, 2)\n",
      "      gene  proteinrpna\n",
      "0    YWHAB     0.032595\n",
      "1    YWHAE     0.148846\n",
      "2    YWHAZ     0.190503\n",
      "3  TP53BP1     0.787472\n",
      "4     ARAF    -0.070401\n",
      "getProteinRPPA  (19989, 9)\n",
      "no sanger\n",
      "(43607, 3)\n",
      "gc content : (43607, 2)\n",
      "      gene  percentage_gene_gc_content\n",
      "0    MT-TF                       40.85\n",
      "1  MT-RNR1                       45.49\n",
      "2    MT-TV                       42.03\n",
      "3  MT-RNR2                       42.81\n",
      "4   MT-TL1                       38.67\n",
      "getGCContent  (21560, 10)\n",
      "GO Ids: (483081, 2)\n",
      "      gene       go_id\n",
      "0    MT-TF  GO:0030533\n",
      "1    MT-TF  GO:0006412\n",
      "4  MT-RNR2  GO:0003735\n",
      "5  MT-RNR2  GO:0005840\n",
      "6   MT-TL1  GO:0030533\n",
      "getGOgenes  (21560, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>rnaexp</th>\n",
       "      <th>copynumber</th>\n",
       "      <th>crisprcas9</th>\n",
       "      <th>rnai</th>\n",
       "      <th>ccleexp</th>\n",
       "      <th>methylation</th>\n",
       "      <th>proteingygi</th>\n",
       "      <th>proteinrpna</th>\n",
       "      <th>percentage_gene_gc_content</th>\n",
       "      <th>go_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>3.475085</td>\n",
       "      <td>0.846132</td>\n",
       "      <td>-0.064129</td>\n",
       "      <td>0.003667</td>\n",
       "      <td>4.221877</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.582833</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.40</td>\n",
       "      <td>GO:0016020 GO:0005515 GO:0043123 GO:0070062 GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.846132</td>\n",
       "      <td>0.088048</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.78</td>\n",
       "      <td>GO:0005634 GO:0005737 GO:0016020 GO:0005515 GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>5.632850</td>\n",
       "      <td>0.846061</td>\n",
       "      <td>-0.031131</td>\n",
       "      <td>0.297415</td>\n",
       "      <td>5.512227</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.219692</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.20</td>\n",
       "      <td>GO:0005783 GO:0016740 GO:0016757 GO:0006486 GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1.786596</td>\n",
       "      <td>1.069042</td>\n",
       "      <td>-0.090572</td>\n",
       "      <td>-0.166058</td>\n",
       "      <td>1.794936</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.377583</td>\n",
       "      <td>0.0</td>\n",
       "      <td>40.14</td>\n",
       "      <td>GO:0005524 GO:0006468 GO:0004672 GO:0005737 GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FIRRM</td>\n",
       "      <td>2.797013</td>\n",
       "      <td>1.069042</td>\n",
       "      <td>0.026379</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>39.22</td>\n",
       "      <td>GO:0005634 GO:0005634 GO:0005737 GO:0005856 GO...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene    rnaexp  copynumber  crisprcas9      rnai   ccleexp  methylation  \\\n",
       "0  TSPAN6  3.475085    0.846132   -0.064129  0.003667  4.221877          0.0   \n",
       "1    TNMD  0.000000    0.846132    0.088048  0.000000  0.000000          0.0   \n",
       "2    DPM1  5.632850    0.846061   -0.031131  0.297415  5.512227          0.0   \n",
       "3   SCYL3  1.786596    1.069042   -0.090572 -0.166058  1.794936          0.0   \n",
       "4   FIRRM  2.797013    1.069042    0.026379  0.000000  0.000000          0.0   \n",
       "\n",
       "   proteingygi  proteinrpna  percentage_gene_gc_content  \\\n",
       "0    -0.582833          0.0                       40.40   \n",
       "1     0.000000          0.0                       40.78   \n",
       "2    -0.219692          0.0                       40.20   \n",
       "3     0.377583          0.0                       40.14   \n",
       "4     0.000000          0.0                       39.22   \n",
       "\n",
       "                                               go_id  \n",
       "0  GO:0016020 GO:0005515 GO:0043123 GO:0070062 GO...  \n",
       "1  GO:0005634 GO:0005737 GO:0016020 GO:0005515 GO...  \n",
       "2  GO:0005783 GO:0016740 GO:0016757 GO:0006486 GO...  \n",
       "3  GO:0005524 GO:0006468 GO:0004672 GO:0005737 GO...  \n",
       "4  GO:0005634 GO:0005634 GO:0005737 GO:0005856 GO...  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final = getRNAExpression(CELLLINE)\n",
    "df_final.drop_duplicates(subset='gene', keep='first')\n",
    "print('getRNAExpression ',df_final.shape)\n",
    "\n",
    "dftemp1 = getCN(CELLLINE)\n",
    "dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "df_final = df_final.fillna(0)\n",
    "print('getCN ',df_final.shape)\n",
    "\n",
    "dftemp1 = getCRISPRCas9(CELLLINE)\n",
    "dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "df_final = df_final.fillna(0)\n",
    "print('getCRISPRCas9 ',df_final.shape)\n",
    "\n",
    "\n",
    "dftemp1 = getRNAi(CELLLINE)\n",
    "dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "df_final = df_final.fillna(0)\n",
    "print('getRNAi ',df_final.shape)\n",
    "\n",
    "dftemp1 = getCCLExpression(CELLLINE)\n",
    "dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "df_final = df_final.fillna(0)\n",
    "print('getCCLExpression ',df_final.shape)\n",
    "\n",
    "dftemp1 = getMethylation(CELLLINE)\n",
    "dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "df_final = df_final.fillna(0)\n",
    "print('getMethylation ',df_final.shape)\n",
    "\n",
    "try:\n",
    "    dftemp1 = getProteinGygi(CELLLINE)\n",
    "    dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "    df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "    df_final = df_final.fillna(0)\n",
    "    print('getProteinGygi ',df_final.shape)\n",
    "except:\n",
    "    df_final['proteingygi']=0\n",
    "    \n",
    "try:\n",
    "    dftemp1 = getProteinRPPA(CELLLINE)\n",
    "    dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "    df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "    df_final = df_final.fillna(0)\n",
    "    print('getProteinRPPA ',df_final.shape)\n",
    "except:\n",
    "    df_final['proteinrpna']=0\n",
    "\n",
    "\n",
    "try:\n",
    "    dftemp1 = getSangerCFGs(CELLLINE)\n",
    "    dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "    df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "    df_final = df_final.fillna(0)\n",
    "    print('getSangerCFGs ',df_final.shape)\n",
    "except:\n",
    "    print(\"no sanger\")\n",
    "\n",
    "\n",
    "\n",
    "dftemp1 = getGCContent()\n",
    "dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "df_final = df_final.fillna(0)\n",
    "print('getGCContent ',df_final.shape)\n",
    "\n",
    "\n",
    "dftemp1 = getGOgenes()\n",
    "dftemp1 = dftemp1.groupby('gene')['go_id'].agg(' '.join).reset_index()\n",
    "dftemp1.drop_duplicates(subset='gene', keep='first')\n",
    "df_final = pd.merge(df_final, dftemp1, how='left', on='gene')\n",
    "df_final = df_final.fillna(0)\n",
    "print('getGOgenes ',df_final.shape)\n",
    "\n",
    "df_final.to_csv(CELLINENAME+'.csv',index=False)\n",
    "df_final.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8379a3d7-4988-421a-a95f-bbf9a1f7a208",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83e82c79-2f9c-4db8-83e0-9a32ce61bbc3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51b6b927-1e11-4561-8ed9-04a11b094cb9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90a9bd51-0bac-41c8-9119-8af3146dc6b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfgo = getGOgenes()\n",
    "dfgo = dfgo.groupby('gene')['go_id'].agg(' '.join).reset_index()\n",
    "\n",
    "dfgo=dfgo.head(10)\n",
    "print(dfgo)\n",
    "G = nx.Graph()\n",
    "for _, row in dfgo.iterrows():\n",
    "    gene = row['gene']\n",
    "    gos=row['go_id'].split(' ')\n",
    "    for go_id in gos:\n",
    "        G.add_edge(gene,go_id,)  \n",
    " \n",
    "plt.figure(figsize=(6, 4))\n",
    "pos = nx.kamada_kawai_layout(G)  \n",
    "nx.draw(G, pos, with_labels=True, node_size=500, node_color='pink', font_size=10, edge_color='darkblue')\n",
    "plt.title(\"Go term - Gene Network\")\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71a01995-8525-4cc8-8ff1-2021bb5b731f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfgo=dfgo.head(200)\n",
    "G = nx.Graph()\n",
    "for _, row in dfgo.iterrows():\n",
    "    gene = row['gene_name']\n",
    "    go_id = row['go_id']\n",
    "    G.add_edge(gene,go_id,)  # Adding an edge between the gene and its associated GO term\n",
    " \n",
    " \n",
    "plt.figure(figsize=(6, 4))\n",
    "pos = nx.kamada_kawai_layout(G)  # Positions for nodes\n",
    "\n",
    "nx.draw(G, pos, with_labels=True, node_size=500, node_color='pink', font_size=10, edge_color='darkblue')\n",
    "plt.title(\"Go term - Gene Network\")\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808929ab-18e9-453b-a025-585e293b13e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "G = nx.erdos_renyi_graph(10,0.3)\n",
    "for _, row in dfgo.iterrows():\n",
    "    gene = row['gene_name']\n",
    "    go_id = row['go_id']\n",
    "    G.add_edge(gene, go_id)  # Adding an edge between the gene and its associated GO term\n",
    " \n",
    "pos = nx.kamada_kawai_layout(G)\n",
    "plt.figure(figsize=(8, 6))\n",
    "nx.draw(G, pos, with_labels=True, node_size=700, node_color='lightblue', font_size=12)\n",
    "plt.title(\"Circular Layout Example\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "833c87bb-68b3-45d6-8b22-02cc0645f7ad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f055628-15b5-4930-91d0-798877c6af41",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7573245-bca4-4805-a995-4c0d42d35263",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
