{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b0c5bbf1-7779-4fd1-ad8f-28b2fd39e183",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install pyranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cd014030-57e2-47cb-bfc5-e793c7d18b14",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd2e5ab4-99ff-4bfd-ab03-530d0905eb6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000456328.2:1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000456328.2:2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000456328.2:3</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000450305.2:1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon:ENST00000450305.2:2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>...</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>PGO:0000005,PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------+\n",
       "| Chromosome   | Source     | Feature    |     Start |       End | Score      | Strand       | Frame      | +19   |\n",
       "| (category)   | (object)   | (object)   |   (int64) |   (int64) | (object)   | (category)   | (object)   | ...   |\n",
       "|--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------|\n",
       "| chr1         | HAVANA     | gene       |     11868 |     14409 | .          | +            | .          | ...   |\n",
       "| chr1         | HAVANA     | transcript |     11868 |     14409 | .          | +            | .          | ...   |\n",
       "| chr1         | HAVANA     | exon       |     11868 |     12227 | .          | +            | .          | ...   |\n",
       "| chr1         | HAVANA     | exon       |     12612 |     12721 | .          | +            | .          | ...   |\n",
       "| chr1         | HAVANA     | exon       |     13220 |     14409 | .          | +            | .          | ...   |\n",
       "| chr1         | HAVANA     | transcript |     12009 |     13670 | .          | +            | .          | ...   |\n",
       "| chr1         | HAVANA     | exon       |     12009 |     12057 | .          | +            | .          | ...   |\n",
       "| chr1         | HAVANA     | exon       |     12178 |     12227 | .          | +            | .          | ...   |\n",
       "+--------------+------------+------------+-----------+-----------+------------+--------------+------------+-------+\n",
       "Stranded PyRanges object has 8 rows and 27 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "19 hidden columns: ID, gene_id, gene_type, gene_name, level, hgnc_id, havana_gene, Parent, transcript_id, transcript_type, ... (+ 9 more.)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation = pr.read_gff3(\"gene.gff3\") \n",
    "annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4920ee1c-0b4c-4a18-bac9-5067ee275da6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+-------------+\n",
       "| Chromosome   |     Start |       End | Strand       | gene_name   |\n",
       "| (category)   |   (int64) |   (int64) | (category)   | (object)    |\n",
       "|--------------+-----------+-----------+--------------+-------------|\n",
       "| chr1         |     11868 |     14409 | +            | DDX11L1     |\n",
       "| chr1         |     11868 |     14409 | +            | DDX11L1     |\n",
       "| chr1         |     11868 |     12227 | +            | DDX11L1     |\n",
       "| chr1         |     12612 |     12721 | +            | DDX11L1     |\n",
       "| chr1         |     13220 |     14409 | +            | DDX11L1     |\n",
       "| chr1         |     12009 |     13670 | +            | DDX11L1     |\n",
       "| chr1         |     12009 |     12057 | +            | DDX11L1     |\n",
       "| chr1         |     12178 |     12227 | +            | DDX11L1     |\n",
       "+--------------+-----------+-----------+--------------+-------------+\n",
       "Stranded PyRanges object has 8 rows and 5 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand."
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation[['gene_name']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "508cec98-c1f2-4b16-b800-b815ab4f15db",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e56be89d-0e50-415f-8954-971f512ce8f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "breast_ATAC-seq.bed\n",
      "breast_ATAC-seq\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "| Chromosome   |     Start |       End | Name       |     Score | Strand       |   ThickStart |    ThickEnd | +2    |\n",
      "| (category)   |   (int64) |   (int64) | (object)   |   (int64) | (category)   |    (float64) |   (float64) | ...   |\n",
      "|--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------|\n",
      "| chr1         | 100037824 | 100038965 | Peak_37655 |       230 | .            |      4.28616 |     23.0381 | ...   |\n",
      "| chr1         | 100037824 | 100038965 | Peak_38635 |       224 | .            |      4.23113 |     22.422  | ...   |\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "Unstranded PyRanges object has 2 rows and 10 columns from 24 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "Considered unstranded due to these Strand values: '.'\n",
      "2 hidden columns: ItemRGB, BlockCount\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand', 'ThickStart',\n",
      "       'ThickEnd', 'ItemRGB', 'BlockCount', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score_b', 'Strand_b', 'Frame', 'ID', 'gene_id', 'gene_type',\n",
      "       'gene_name', 'level', 'hgnc_id', 'havana_gene', 'Parent',\n",
      "       'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "              gene  breast_ATAC-seq\n",
      "0  ENSG00000283761              230\n",
      "9  ENSG00000288826              230\n",
      "breast_chipreq_H3K4me1.bed\n",
      "breast_chipreq_H3K4me1\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "| Chromosome   |     Start |       End | Name       |     Score | Strand       |   ThickStart |    ThickEnd | +2    |\n",
      "| (category)   |   (int64) |   (int64) | (object)   |   (int64) | (category)   |    (float64) |   (float64) | ...   |\n",
      "|--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------|\n",
      "| chr1         | 100036820 | 100037564 | Peak_18350 |        67 | .            |      3.29161 |     6.79591 | ...   |\n",
      "| chr1         | 100248433 | 100249356 | Peak_2278  |       105 | .            |      4.34989 |    10.5579  | ...   |\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "Unstranded PyRanges object has 2 rows and 10 columns from 23 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "Considered unstranded due to these Strand values: '.'\n",
      "2 hidden columns: ItemRGB, BlockCount\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand', 'ThickStart',\n",
      "       'ThickEnd', 'ItemRGB', 'BlockCount', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score_b', 'Strand_b', 'Frame', 'ID', 'gene_id', 'gene_type',\n",
      "       'gene_name', 'level', 'hgnc_id', 'havana_gene', 'Parent',\n",
      "       'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "              gene  breast_chipreq_H3K4me1\n",
      "0  ENSG00000283761                      67\n",
      "9  ENSG00000288826                      67\n",
      "breast_chipreq_H3K4me3.bed\n",
      "breast_chipreq_H3K4me3\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "| Chromosome   |     Start |       End | Name       |     Score | Strand       |   ThickStart |    ThickEnd | +2    |\n",
      "| (category)   |   (int64) |   (int64) | (object)   |   (int64) | (category)   |    (float64) |   (float64) | ...   |\n",
      "|--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------|\n",
      "| chr1         | 100037047 | 100039662 | Peak_638   |       780 | .            |      68.2588 |     504.063 | ...   |\n",
      "| chr1         | 100132135 | 100134184 | Peak_5517  |       535 | .            |      58.1115 |     344.435 | ...   |\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "Unstranded PyRanges object has 2 rows and 10 columns from 32 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "Considered unstranded due to these Strand values: '.'\n",
      "2 hidden columns: ItemRGB, BlockCount\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand', 'ThickStart',\n",
      "       'ThickEnd', 'ItemRGB', 'BlockCount', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score_b', 'Strand_b', 'Frame', 'ID', 'gene_id', 'gene_type',\n",
      "       'gene_name', 'level', 'hgnc_id', 'havana_gene', 'Parent',\n",
      "       'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "              gene  breast_chipreq_H3K4me3\n",
      "0  ENSG00000283761                     780\n",
      "9  ENSG00000288826                     780\n",
      "breast_chipreq_H3K9me3.bed\n",
      "breast_chipreq_H3K9me3\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "| Chromosome   |     Start |       End | Name       |     Score | Strand       |   ThickStart |    ThickEnd | +2    |\n",
      "| (category)   |   (int64) |   (int64) | (object)   |   (int64) | (category)   |    (float64) |   (float64) | ...   |\n",
      "|--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------|\n",
      "| chr1         | 100204617 | 100205354 | Peak_18972 |        97 | .            |      3.91906 |     9.70126 | ...   |\n",
      "| chr1         | 100315679 | 100316723 | Peak_34799 |        78 | .            |      3.4296  |     7.83805 | ...   |\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "Unstranded PyRanges object has 2 rows and 10 columns from 23 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "Considered unstranded due to these Strand values: '.'\n",
      "2 hidden columns: ItemRGB, BlockCount\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand', 'ThickStart',\n",
      "       'ThickEnd', 'ItemRGB', 'BlockCount', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score_b', 'Strand_b', 'Frame', 'ID', 'gene_id', 'gene_type',\n",
      "       'gene_name', 'level', 'hgnc_id', 'havana_gene', 'Parent',\n",
      "       'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "    gene  breast_chipreq_H3K9me3\n",
      "0    DBT                      97\n",
      "4  EXTL2                     121\n",
      "breast_DNase-seq.bed\n",
      "breast_DNase-seq\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+------------+-------+\n",
      "| Chromosome   |     Start |       End | Name       |     Score | Strand       |   ThickStart |   ThickEnd | +2    |\n",
      "| (category)   |   (int64) |   (int64) | (object)   |   (int64) | (category)   |    (float64) |    (int64) | ...   |\n",
      "|--------------+-----------+-----------+------------+-----------+--------------+--------------+------------+-------|\n",
      "| chr1         |     10452 |     10620 | .          |         0 | .            |     0.324654 |         -1 | ...   |\n",
      "| chr1         |     17420 |     17560 | .          |         0 | .            |     0.505017 |         -1 | ...   |\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+------------+-------+\n",
      "Unstranded PyRanges object has 2 rows and 10 columns from 24 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "Considered unstranded due to these Strand values: '.'\n",
      "2 hidden columns: ItemRGB, BlockCount\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand', 'ThickStart',\n",
      "       'ThickEnd', 'ItemRGB', 'BlockCount', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score_b', 'Strand_b', 'Frame', 'ID', 'gene_id', 'gene_type',\n",
      "       'gene_name', 'level', 'hgnc_id', 'havana_gene', 'Parent',\n",
      "       'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "        gene  breast_DNase-seq\n",
      "0     WASH7P                 0\n",
      "2  MIR6859-1                 0\n",
      "breast_tfchipseq_CTCF.bed\n",
      "breast_tfchipseq_CTCF\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "| Chromosome   |     Start |       End | Name       |     Score | Strand       |   ThickStart |    ThickEnd | +2    |\n",
      "| (category)   |   (int64) |   (int64) | (object)   |   (int64) | (category)   |    (float64) |   (float64) | ...   |\n",
      "|--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------|\n",
      "| chr1         |  23019568 |  23019972 | .          |       992 | .            |      7.34306 |          -1 | ...   |\n",
      "| chr1         | 223708259 | 223708663 | .          |       704 | .            |      7.34865 |          -1 | ...   |\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "Unstranded PyRanges object has 2 rows and 10 columns from 35 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "Considered unstranded due to these Strand values: '.'\n",
      "2 hidden columns: ItemRGB, BlockCount\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand', 'ThickStart',\n",
      "       'ThickEnd', 'ItemRGB', 'BlockCount', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score_b', 'Strand_b', 'Frame', 'ID', 'gene_id', 'gene_type',\n",
      "       'gene_name', 'level', 'hgnc_id', 'havana_gene', 'Parent',\n",
      "       'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "      gene  breast_tfchipseq_CTCF\n",
      "0    KDM1A                    992\n",
      "116  CAPN2                    704\n",
      "breast_tfchipseq_POLR2A.bed\n",
      "breast_tfchipseq_POLR2A\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "| Chromosome   |     Start |       End | Name       |     Score | Strand       |   ThickStart |    ThickEnd | +2    |\n",
      "| (category)   |   (int64) |   (int64) | (object)   |   (int64) | (category)   |    (float64) |   (float64) | ...   |\n",
      "|--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------|\n",
      "| chr1         | 156737369 | 156737805 | .          |      1000 | .            |      5.2861  |          -1 | ...   |\n",
      "| chr1         | 154580764 | 154581200 | .          |       841 | .            |      5.29237 |          -1 | ...   |\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "Unstranded PyRanges object has 2 rows and 10 columns from 23 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "Considered unstranded due to these Strand values: '.'\n",
      "2 hidden columns: ItemRGB, BlockCount\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand', 'ThickStart',\n",
      "       'ThickEnd', 'ItemRGB', 'BlockCount', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score_b', 'Strand_b', 'Frame', 'ID', 'gene_id', 'gene_type',\n",
      "       'gene_name', 'level', 'hgnc_id', 'havana_gene', 'Parent',\n",
      "       'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "      gene  breast_tfchipseq_POLR2A\n",
      "0   MRPL24                     1000\n",
      "23   TIPRL                     1000\n",
      "breast_tfchipseq_POLR2AphosphoS5.bed\n",
      "breast_tfchipseq_POLR2AphosphoS5\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "| Chromosome   |     Start |       End | Name       |     Score | Strand       |   ThickStart |    ThickEnd | +2    |\n",
      "| (category)   |   (int64) |   (int64) | (object)   |   (int64) | (category)   |    (float64) |   (float64) | ...   |\n",
      "|--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------|\n",
      "| chr1         | 153990560 | 153991041 | .          |      1000 | .            |      663.128 |          -1 | ...   |\n",
      "| chr1         |  28648421 |  28648806 | .          |      1000 | .            |      618.765 |          -1 | ...   |\n",
      "+--------------+-----------+-----------+------------+-----------+--------------+--------------+-------------+-------+\n",
      "Unstranded PyRanges object has 2 rows and 10 columns from 24 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "Considered unstranded due to these Strand values: '.'\n",
      "2 hidden columns: ItemRGB, BlockCount\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "join: Strand data from other will be added as strand data to self.\n",
      "If this is undesired use the flag apply_strand_suffix=False.\n",
      "To turn off the warning set apply_strand_suffix to True or False.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Start', 'End', 'Name', 'Score', 'Strand', 'ThickStart',\n",
      "       'ThickEnd', 'ItemRGB', 'BlockCount', 'Source', 'Feature', 'Start_b',\n",
      "       'End_b', 'Score_b', 'Strand_b', 'Frame', 'ID', 'gene_id', 'gene_type',\n",
      "       'gene_name', 'level', 'hgnc_id', 'havana_gene', 'Parent',\n",
      "       'transcript_id', 'transcript_type', 'transcript_name',\n",
      "       'transcript_support_level', 'tag', 'havana_transcript', 'exon_number',\n",
      "       'exon_id', 'ont', 'protein_id', 'ccdsid'],\n",
      "      dtype='object')\n",
      "        gene  breast_tfchipseq_POLR2AphosphoS5\n",
      "0      RPS27                              1000\n",
      "19  TAF12-DT                              1000\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "bed_files = glob.glob('*.bed')\n",
    "for file in bed_files:\n",
    "    try :\n",
    "        print(file)\n",
    "        featname = os.path.splitext(file)[0]  \n",
    "        print(featname)\n",
    "        peaks = pr.read_bed(file)\n",
    "        print(peaks.head(2))\n",
    "        overlapping_genes = peaks.join(annotation, how=\"left\") \n",
    "        print(overlapping_genes.columns)\n",
    "        \n",
    "        df_overlapping_genes = overlapping_genes.df\n",
    "        df=df_overlapping_genes[['gene_name','Score']].drop_duplicates()\n",
    "        df = df.rename(columns={'gene_name': 'gene'})\n",
    "        df = df.rename(columns={'Score':featname})\n",
    "        print(df.head(2))\n",
    "        df.to_csv(featname+'.csv',index=False)\n",
    "    except:\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de4d9d62-2306-4971-b061-9747cbb811e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba0e2dc7-fd20-4cc7-ad24-b35aed4468e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fbfe7560-f04f-48b6-a714-ccec070ac45c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(120498, 2)\n",
      "(19884, 2)\n",
      "(27010, 2)\n",
      "(10379, 2)\n",
      "\n",
      "1 (19884, 3)\n",
      "(30204, 2)\n",
      "(17947, 2)\n",
      "\n",
      "2 (19884, 4)\n",
      "(33377, 2)\n",
      "(6752, 2)\n",
      "\n",
      "3 (19884, 5)\n",
      "(19825, 2)\n",
      "(16087, 2)\n",
      "\n",
      "4 (19884, 6)\n",
      "(19518, 2)\n",
      "(11685, 2)\n",
      "\n",
      "5 (19884, 7)\n",
      "(9247, 2)\n",
      "(5208, 2)\n",
      "\n",
      "6 (19884, 8)\n",
      "(25326, 2)\n",
      "(12320, 2)\n",
      "\n",
      "7 (19884, 9)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "\n",
    "csv_files = glob.glob('*.csv')\n",
    "merged_df = pd.DataFrame()\n",
    "for i, file in enumerate(csv_files):\n",
    "    current_df = pd.read_csv(file)\n",
    "    print(current_df.shape)\n",
    "    current_df = current_df[~current_df['gene'].str.startswith('ENSG0')].drop_duplicates(subset='gene')\n",
    "    print(current_df.shape)\n",
    "    if i == 0:\n",
    "        merged_df = current_df\n",
    "    else:\n",
    "        print(\"\")\n",
    "        merged_df = pd.merge(merged_df, current_df, on='gene', how='left')  # 'outer' \n",
    "        print(i,merged_df.shape)\n",
    "merged_df.to_csv('BREAST_CHIPSEQ.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f743475-2f47-445f-b970-e14970fda3c3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
